source/extensions/local_address_selectors/filter_state_override/config.cc

```diff
@@ -0,0 +1,71 @@
+#include "source/extensions/local_address_selectors/filter_state_override/config.h"
+
+#include "envoy/registry/registry.h"
+#include "envoy/upstream/upstream.h"
+
+#include "source/common/upstream/default_local_address_selector_factory.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace LocalAddressSelectors {
+namespace FilterStateOverride {
+
+namespace {
+class NamespaceLocalAddressSelector : public Upstream::UpstreamLocalAddressSelector,
+                                      public Logger::Loggable<Logger::Id::upstream> {
+public:
+  explicit NamespaceLocalAddressSelector(
+      const Upstream::UpstreamLocalAddressSelectorConstSharedPtr& inner)
+      : inner_(inner) {}
+  Upstream::UpstreamLocalAddress getUpstreamLocalAddress(
+      const Network::Address::InstanceConstSharedPtr& endpoint_address,
+      const Network::ConnectionSocket::OptionsSharedPtr& socket_options,
+      OptRef<const Network::TransportSocketOptions> transport_socket_options) const {
+    const auto upstream_address =
+        inner_->getUpstreamLocalAddress(endpoint_address, socket_options, transport_socket_options);
+    if (transport_socket_options && upstream_address.address_) {
+      for (const auto& obj : transport_socket_options->downstreamSharedFilterStateObjects()) {
+        if (obj.name_ == "envoy.network.upstream_network_namespace") {
+          if (const auto data = obj.data_->serializeAsString(); data.has_value()) {
+            const auto new_address = upstream_address.address_->withNetworkNamespace(*data);
+            if (new_address) {
+              return {.address_ = new_address, .socket_options_ = upstream_address.socket_options_};
+            }
+          } else {
+            ENVOY_LOG(trace, "Failed to serialize filter state as string");
+          }
+          break;
+        }
+      }
+    }
```

Comment:
Can you make some helper functions to reduce the amount of nesting? Would make this easier to read.

Reply:
Done.